locator/abstract_replication_strategy.cc

```diff
@@ -256,18 +256,9 @@ abstract_replication_strategy::get_address_ranges(const token_metadata& tm, inet
     co_return ret;
 }
 
-std::unordered_map<dht::token_range, inet_address_vector_replica_set>
+future<std::unordered_map<dht::token_range, inet_address_vector_replica_set>>
 effective_replication_map::get_range_addresses() const {
-    const token_metadata& tm = *_tmptr;
-    std::unordered_map<dht::token_range, inet_address_vector_replica_set> ret;
-    for (auto& t : tm.sorted_tokens()) {
-        dht::token_range_vector ranges = tm.get_primary_ranges_for(t);
-        auto eps = get_natural_endpoints(t);
-        for (auto& r : ranges) {
-            ret.emplace(r, eps);
-        }
-    }
-    return ret;
+    return _rs->get_range_addresses(*_tmptr);
```

Comment:
1. Can it be precomputed? Probably wasteful since it's rarely needed
2. Can it be optimized by merging precomputed results of get_natural_endpoint()? Probably we do have get_natural_endpoints(t) cached.
3. Can it be improved for everywhere_replication_strategy by making its effective_replication_map single-range (ignoring vnodes)? Probably

Reply:
> 1. Can it be precomputed? Probably wasteful since it's rarely needed

Agreed.

> 2. Can it be optimized by merging precomputed results of get_natural_endpoint()? Probably we do have get_natural_endpoints(t) cached.

Hmm, yes.  Good catch.
We can use the replication_map directly, as `get_natural_endpoints` simply searches the replication map
or calls the replication strategy specific api.  The replication map is already pre-calculated with the natural endpoint so we should be able to safely use it.

> 3. Can it be improved for everywhere_replication_strategy by making its effective_replication_map single-range (ignoring vnodes)? Probably

Agreed. but that's a far bigger change.